File: client/lib/treasure-detector.ts
import {
  MagneticReading,
  TreasureResult,
  ResourceType,
  RESOURCE_THRESHOLDS,
} from "@shared/magnetometer";

export class TreasureDetector {
  private readonly anomalyThreshold: number = 15; // Anomali eşiği (µT)

  /**
   * Manyetik ölçümlerden kaynakları tespit et
   */
  detectResources(
    readings: MagneticReading[],
    baselineStrength: number
  ): TreasureResult[] {
    const results: TreasureResult[] = [];
    let anomalyCount = 0;

    readings.forEach((reading, index) => {
      const anomalyLevel = Math.abs(reading.total - baselineStrength);

      // Anomali kontrolü
      if (anomalyLevel > this.anomalyThreshold) {
        anomalyCount++;

        const resourceType = this.classifyResource(reading.total);
        const confidence = this.calculateConfidence(
          anomalyLevel,
          resourceType,
          anomalyCount
        );

        if (confidence > 30) {
          // Minimum %30 güven
          results.push({
            id: `treasure_${Date.now()}_${index}`,
            resourceType,
            magneticStrength: reading.total,
            anomalyLevel,
            timestamp: reading.timestamp,
            confidence,
          });
        }
      }
    });

    return results;
  }

  /**
   * Manyetik şiddete göre kaynak türünü sınıflandır
   */
  private classifyResource(magneticStrength: number): ResourceType {
    if (magneticStrength >= RESOURCE_THRESHOLDS[ResourceType.ARTIFACT]) {
      return ResourceType.ARTIFACT;
    } else if (
      magneticStrength >= RESOURCE_THRESHOLDS[ResourceType.VALUABLE_MATERIAL]
    ) {
      return ResourceType.VALUABLE_MATERIAL;
    } else if (magneticStrength >= RESOURCE_THRESHOLDS[ResourceType.MINERAL]) {
      return ResourceType.MINERAL;
    } else if (
      magneticStrength >=
      RESOURCE_THRESHOLDS[ResourceType.UNDERGROUND_STRUCTURE]
    ) {
      return ResourceType.UNDERGROUND_STRUCTURE;
    }
    return ResourceType.UNKNOWN;
  }

  /**
   * Kaynak güven düzeyini hesapla (0-100)
   */
  private calculateConfidence(
    anomalyLevel: number,
    resourceType: ResourceType,
    detectionCount: number
  ): number {
    let confidence = 50; // Başlangıç güveni

    // Anomali seviyesine göre arttır
    confidence += Math.min(anomalyLevel * 2, 30);

    // Kaynak türüne göre ayarla
    const typeConfidence: Record<string, number> = {
      [ResourceType.ARTIFACT]: 20,
      [ResourceType.VALUABLE_MATERIAL]: 15,
      [ResourceType.MINERAL]: 10,
      [ResourceType.UNDERGROUND_STRUCTURE]: 5,
      [ResourceType.UNKNOWN]: -10,
    };

    confidence += typeConfidence[resourceType] || 0;

    // Tekrarlayan deteksiyonlar güveni arttırır
    confidence += Math.min(detectionCount * 2, 15);

    return Math.min(Math.max(confidence, 0), 100);
  }

  /**
   * Eşiği ayarla
   */
  setAnomalyThreshold(threshold: number): void {
    if (threshold > 0 && threshold < 100) {
      (this.anomalyThreshold as any) = threshold;
    }
  }

  /**
   * Önceki ölçümlerden trend analizi yap
   */
  analyzeTrend(readings: MagneticReading[]): {
    trend: "increasing" | "decreasing" | "stable";
    average: number;
    max: number;
    min: number;
  } {
    if (readings.length === 0) {
      return {
        trend: "stable",
        average: 0,
        max: 0,
        min: 0,
      };
    }

    const values = readings.map((r) => r.total);
    const average = values.reduce((a, b) => a + b, 0) / values.length;
    const max = Math.max(...values);
    const min = Math.min(...values);

    // Son 5 okumaya bakarak trend belirle
    const recentReadings = readings.slice(-5);
    const recentValues = recentReadings.map((r) => r.total);

    let trend: "increasing" | "decreasing" | "stable" = "stable";
    if (recentValues.length > 1) {
      const slope =
        (recentValues[recentValues.length - 1] - recentValues[0]) /
        recentValues.length;
      if (slope > 2) {
        trend = "increasing";
      } else if (slope < -2) {
        trend = "decreasing";
      }
    }

    return {
      trend,
      average: parseFloat(average.toFixed(2)),
      max: parseFloat(max.toFixed(2)),
      min: parseFloat(min.toFixed(2)),
    };
  }
}

File: ANDROID_PROJECT/app/src/main/java/com/example/magnetictreasure/detection/TreasureDetector.kt
package com.example.magnetictreasure.detection

import com.example.magnetictreasure.sensor.MagneticReading
import kotlin.math.abs

enum class ResourceType {
    TREASURE,                    // 💎 >120 µT
    VALUABLE_MATERIAL,           // ✨ >110 µT
    MINERAL,                     // ⛏️ >95 µT
    UNDERGROUND_STRUCTURE,       // 🏗️ >80 µT
    UNKNOWN                      // ❓
}

data class TreasureResult(
    val id: String = java.util.UUID.randomUUID().toString(),
    val resourceType: ResourceType,
    val magneticStrength: Float,
    val anomalyLevel: Float,
    val confidence: Float, // 0-100
    val latitude: Double? = null,
    val longitude: Double? = null,
    val timestamp: Long = System.currentTimeMillis()
)

class TreasureDetector {
    
    private val anomalyThreshold: Float = 15f // µT
    
    private val thresholds = mapOf(
        ResourceType.TREASURE to 120f,
        ResourceType.VALUABLE_MATERIAL to 110f,
        ResourceType.MINERAL to 95f,
        ResourceType.UNDERGROUND_STRUCTURE to 80f,
        ResourceType.UNKNOWN to 0f
    )
    
    fun detectResources(
        readings: List<MagneticReading>,
        baselineStrength: Float
    ): List<TreasureResult> {
        val results = mutableListOf<TreasureResult>()
        var anomalyCount = 0
        
        readings.forEach { reading ->
            val anomalyLevel = abs(reading.total - baselineStrength)
            
            if (anomalyLevel > anomalyThreshold) {
                anomalyCount++
                
                val resourceType = classifyResource(reading.total)
                val confidence = calculateConfidence(
                    anomalyLevel, 
                    resourceType, 
                    anomalyCount
                )
                
                if (confidence > 30) {
                    results.add(TreasureResult(
                        resourceType = resourceType,
                        magneticStrength = reading.total,
                        anomalyLevel = anomalyLevel,
                        confidence = confidence,
                        timestamp = reading.timestamp
                    ))
                }
            }
        }
        
        return results
    }
    
    private fun classifyResource(magneticStrength: Float): ResourceType {
        return when {
            magneticStrength >= thresholds[ResourceType.TREASURE]!! -> ResourceType.TREASURE
            magneticStrength >= thresholds[ResourceType.VALUABLE_MATERIAL]!! -> ResourceType.VALUABLE_MATERIAL
            magneticStrength >= thresholds[ResourceType.MINERAL]!! -> ResourceType.MINERAL
            magneticStrength >= thresholds[ResourceType.UNDERGROUND_STRUCTURE]!! -> ResourceType.UNDERGROUND_STRUCTURE
            else -> ResourceType.UNKNOWN
        }
    }
    
    private fun calculateConfidence(
        anomalyLevel: Float,
        resourceType: ResourceType,
        detectionCount: Int
    ): Float {
        var confidence = 50f // Başlangıç güveni
        
        // Anomali seviyesine göre arttır
        confidence += minOf(anomalyLevel * 2, 30f)
        
        // Kaynak türüne göre ayarla
        val typeConfidence = when (resourceType) {
            ResourceType.TREASURE -> 20f
            ResourceType.VALUABLE_MATERIAL -> 15f
            ResourceType.MINERAL -> 10f
            ResourceType.UNDERGROUND_STRUCTURE -> 5f
            ResourceType.UNKNOWN -> -10f
        }
        confidence += typeConfidence
        
        // Tekrarlayan deteksiyonlar güveni arttırır
        confidence += minOf(detectionCount * 2, 15f)
        
        return confidence.coerceIn(0f, 100f)
    }
    
    fun getResourceDisplayName(type: ResourceType): String = when (type) {
        ResourceType.TREASURE -> "💎 Hazine"
        ResourceType.VALUABLE_MATERIAL -> "✨ Değerli Materyal"
        ResourceType.MINERAL -> "⛏️ Madeni Kaynak"
        ResourceType.UNDERGROUND_STRUCTURE -> "🏗️ Yer Altı Yapısı"
        ResourceType.UNKNOWN -> "❓ Bilinmeyen"
    }
    
    fun getResourceColor(type: ResourceType): Int = when (type) {
        ResourceType.TREASURE -> 0xFFD700.toInt() // Gold
        ResourceType.VALUABLE_MATERIAL -> 0xFF0000.toInt() // Red
        ResourceType.MINERAL -> 0x00FFFF.toInt() // Cyan
        ResourceType.UNDERGROUND_STRUCTURE -> 0x808080.toInt() // Gray
        ResourceType.UNKNOWN -> 0xFFFFFF.toInt() // White
    }
}
